### DISCUSSION TITLE
Replacing JAX backend with custom autodiff, possibly complex-step?

### DISCUSSION THREAD

**Discussion by jpivarski** [scikit-hep/awkward#3349] [Ideas]
As presented at https://indico.cern.ch/event/1387764/
and discussed in https://iris-hep.slack.com/archives/C0155BGPGE4/p1734127343384079

The idea is to replace Awkward Array's JAX backend with

* nothing (give up on autodiff)
* a mini-library with an API that would be a better backend
* functionality built right into Awkward Array itself.

This is based on the observation that autodiff is not hard to implement, and all of the troubles we're going through to _partially_ implement autodiff through JAX might be more easily solved by a custom autodiff with a friendlier API.

Here's a complete demonstrator of eager, forward-mode autodiff using the complex-step technique:

https://github.com/scikit-hep/awkward/blob/c59a49c0131186634ae269ea3a176ae70c058c23/studies/autodiff/eager_forward.py#L11-L123

which can be called like

```python
>>> x = np.linspace(-20, 20, 10000)
>>> da_x = diffarray(x)
>>> da_y = np.sin(da_x) / da_x
>>> da_x
diffarray([-20.        -19.9959996 -19.9919992 ...  19.9919992  19.9959996
            20.       ],
          [1. 1. 1. ... 1. 1. 1.])
>>> da_y
diffarray([0.04564726 0.04557439 0.04550076 ... 0.04550076 0.04557439 0.04564726],
          [-0.01812174 -0.01831149 -0.01850102 ...  0.01850102  0.01831149
            0.01812174])
>>> abs(da_y.tangent - ((x*np.cos(x) - np.sin(x)) / x**2)).max()
3.9683650809863025e-10
>>> import matplotlib.pyplot as plt
>>> plt.plot(x, da_y.tangent)
>>> plt.plot(x, (x*np.cos(x) - np.sin(x)) / x**2, ls="--")
```

to get

![image](https://github.com/user-attachments/assets/00c831ca-059a-49f9-ad12-84771f610d10)

However, we'll probably need backpropagation, which will require some sort of DAG to reverse the steps in the calculation, and maybe we'd use our typetracers or dask-awkward to make that DAG. (Our typetracers can be used to make a low-level DAG, with one node per _buffer_ operation, whereas a dask-awkward DAG is high-level. Thinking about this now, I think it _has to_ be a low-level DAG.)

I want to ask everyone (@matthewfeickert, @kratsg, @pfackeldey, @alexander-held, @gordonwatts) whether this is sufficient, and generally what your thoughts are on requirements. I'll try answering a few questions that came up in the talk in subsequent comments.

**Comment by jpivarski** [comment]
The reason I didn't take this seriously 4 months ago was because the "[holomorphic](https://en.wikipedia.org/wiki/Holomorphic_function)" requirement for the complex function seemed like a tight constraint that few functions would satisfy. What changed my mind is that regular autodiff blithely ignores not-actually-differentiable points such as $x = 0$ in ReLU, and [that's okay for our purposes](https://hsf-training.github.io/deep-learning-intro-for-hep/11-minimizers.html#derivatives-and-the-activation-function), so maybe the not-actually-correct answers you get with this method are okay, too.

In particular, logarithms have a branch cut where they fail to be holomorphic. What does this method do there?

```python
>>> import numpy as np
>>> from eager_forward import *
>>> x = np.linspace(0, 100, 10000)
>>> da_x = diffarray(x)
>>> da_y = np.log(da_x)
>>> truth = 1/x
<stdin>:1: RuntimeWarning: divide by zero encountered in divide
>>> da_y.tangent
array([1.57079633e+04, 9.99866679e+01, 4.99945835e+01, ...,
       1.00020006e-02, 1.00010002e-02, 1.00000000e-02])
>>> truth
array([           inf, 9.99900000e+01, 4.99950000e+01, ...,
       1.00020006e-02, 1.00010002e-02, 1.00000000e-02])
>>> abs(da_y.tangent - truth)[1:].max()
0.0033321335475875458
>>> abs(da_y.tangent - truth)[1:].argmax()
0
```

Instead of reporting `Inf` or `NaN`, this technique reports a large number. Moreover, the numerical error is maximal near the singularity. Well, that's not surprising, but the errors are close to the singular point.

How about using the logarithm with chain rule?

```python
>>> import numpy as np
>>> import matplotlib.pyplot as plt
>>> from eager_forward import *
>>> x = np.linspace(-10, 10, 10000)
>>> da_x = diffarray(x)
>>> da_y = np.log(np.sin(da_x)**2)
>>> truth = 2*np.sin(x)*np.cos(x) / np.sin(x)**2
>>> da_y.tangent
array([-3.08470206, -3.09826064, -3.11190351, ...,  3.11190351,
        3.09826064,  3.08470206])
>>> truth
array([-3.08470209, -3.09826068, -3.11190355, ...,  3.11190355,
        3.09826068,  3.08470209])
>>> plt.plot(x, da_y.tangent)
[<matplotlib.lines.Line2D object at 0x7d67681b0a50>]
>>> plt.plot(x, truth, ls="--")
[<matplotlib.lines.Line2D object at 0x7d67681b1650>]
>>> plt.show()
```

![image](https://github.com/user-attachments/assets/51e58245-cf05-4e53-a4a4-5bdd508a9f03)

**Reply by gordonwatts** [reply]
Reading a bit about how this works - seems like many frameworks actually do some work to deal with this (as ReLU is quite common).

1. It very often does not hit exactly zero, which is where ReLU is not differentiable.
2. When you do hit exactly zero, it will return 0 or 1 for the slope - so has an artificial "cap"
3. Training with ReLU can be done in smaller batches, etc., so that most batches will report reasonable gradients and smooth out the ones that do not.

Apparently, these are all techniques that are used (according to search engines and GPT's).

**Comment by jpivarski** [comment]
About ReLU'(0), @pfackeldey pointed me to this: https://dl.acm.org/doi/10.5555/3540261.3540297

But from my perspective, the ReLU'(0) case is just what changed my mind about considering this technique that has known errors for functions that aren't perfectly smooth. If we can get away with saying that ReLU'(0) has a value, what else can we get away with? `:)`
